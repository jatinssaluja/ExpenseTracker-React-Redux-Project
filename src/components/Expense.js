import React from 'react';
import {removeExpense} from '../actions/expenses';
import {Link} from 'react-router-dom';
import moment from 'moment';
import numeral from 'numeral';

const Expense = (props)=>(

      <Link className="list-item" to={`/edit/${props.expense.id}`}>

          <div>
             <h3 className="list-item__title">{props.expense.description}</h3>
             <span className="list-item__subtitle">{moment(props.expense.createdAt).format('MMMM Do, YYYY')}</span>
           </div>

        <h3>{numeral(props.expense.amount/100).format('$0,0.00')}</h3>

      </Link>
	);




export default Expense;
