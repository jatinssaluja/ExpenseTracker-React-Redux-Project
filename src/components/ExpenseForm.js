import React, {Component} from 'react';
import moment from 'moment';
import {SingleDatePicker} from 'react-dates';
//import 'react-dates/lib/css/_datepicker.css';

class ExpenseForm extends Component {

	constructor(props) {

      super(props);

      this.onDescriptionChange = this.onDescriptionChange.bind(this);
      this.onNoteChange = this.onNoteChange.bind(this);
      this.onAmountChange = this.onAmountChange.bind(this);
      this.onDateChange = this.onDateChange.bind(this);
      this.onFocusChange = this.onFocusChange.bind(this);
      this.onSubmit = this.onSubmit.bind(this);

      this.state = {

       description: props.expense ? props.expense.description : '',
       note:props.expense ? props.expense.note : '',
       amount:props.expense ? (props.expense.amount/100).toString() : '',
       createdAt: props.expense ? moment(props.expense.createdAt) : moment(),
       calendarFocused:false,
       error:''

      };



	}


   onDescriptionChange(e) {
 
   const description = e.target.value;

    this.setState(()=>{

     return {
  
     description:description

     };


    });


   }


  onNoteChange(e) {

    const note = e.target.value;

    this.setState(()=>{

     return {note};


    });


   }

 onAmountChange(e) {

    const amount = e.target.value;
   
    if(!amount || amount.match(/^\d+(\.\d{0,2})?$/)) {

    this.setState(()=>({amount}));

      }
   }


   onDateChange(createdAt) {

    if(createdAt){

    this.setState(()=>({createdAt}));
     }
      
   }


  onFocusChange({ focused }) {

    this.setState(()=>({calendarFocused:focused}));

      
   }


  onSubmit(e){

   e.preventDefault();

   if(!this.state.description || !this.state.amount){

       this.setState(()=>{

        return {
        
        error:'Please provide description and amount fields'

        };

       });

   }

   else {



      this.props.onSubmit({description:this.state.description, 
        amount:parseFloat(this.state.amount,10)*100, 
        note:this.state.note, createdAt: this.state.createdAt.valueOf() });

      this.setState(()=>{

        return {
        
        error:''

        };

       });

   }


  }





 render() {


 return (

     
      <form className="form" onSubmit = {this.onSubmit}>
        {this.state.error && <p className="form__error">{this.state.error}</p>}

        <input type="text" className = "text-input" placeholder="Description" 
  
         value = {this.state.description}
         onChange = {this.onDescriptionChange}
        autoFocus />
        <input type="text" className = "text-input"

        value = {this.state.amount}
        onChange = {this.onAmountChange}

        placeholder="Amount" />

        <SingleDatePicker

         date = {this.state.createdAt}
         onDateChange = {this.onDateChange}
         focused = {this.state.calendarFocused}
         onFocusChange = {this.onFocusChange}
         numberOfMonths = {1}
         isOutsideRange = {()=>false}

          />



        <textarea className="textarea textarea-mod" placeholder="Add a note for your expense (optional)"

        value = {this.state.note}
        onChange = {this.onNoteChange}
        ></textarea>

        <div>
          <button className="addbtn">Save Expense</button>
          </div>

         </form>

      


 	);


 }



}

export default ExpenseForm;