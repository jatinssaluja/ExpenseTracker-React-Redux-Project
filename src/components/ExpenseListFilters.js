import React, {Component} from 'react';
import {connect} from 'react-redux';
import {setTextFilter, setStartDate, setEndDate} from '../actions/filters';
import {sortByDate, sortByAmount} from '../actions/filters';
import {DateRangePicker} from 'react-dates';

class ExpenseListFilters extends Component {

 
 constructor(props) {

  super(props);

  this.onDatesChange = this.onDatesChange.bind(this);
  this.onFocusChange = this.onFocusChange.bind(this);


  this.state = {

    calendarFocused:null
    


  };





 }


onDatesChange({startDate, endDate}) {


this.props.dispatch(setStartDate(startDate));
this.props.dispatch(setEndDate(endDate));

}


onFocusChange(focused) {

this.setState(()=>({calendarFocused:focused}));


}



 render() {


  return  (<div className="filters-container">


         <div className="input-group">

           <div className="input-group__item">

             <input type="text" className = "text-input" placeholder="Search Expenses" value={this.props.filters.text} onChange={(e)=>{

                this.props.dispatch(setTextFilter(e.target.value));

                }}/>

           </div>

           <div className="input-group__item">
          
              <select className="select" onChange = {(e)=>{

              if(e.target.value === 'date'){
  
                 this.props.dispatch(sortByDate());

                    }

                   else {

                  this.props.dispatch(sortByAmount());

                        }


                    }}>

                 <option value="date">Date</option>
                 <option value="amount">Amount</option>

             </select>
           </div>



           <div className="input-group__item">

              <DateRangePicker

                  startDate={this.props.filters.startDate}
                  endDate = {this.props.filters.endDate}
                  onDatesChange = {this.onDatesChange}
                  focusedInput = {this.state.calendarFocused}
                  onFocusChange = {this.onFocusChange}
                  showClearDates = {true}
                  numberOfMonths = {1}
                  isOutsideRange = {()=>false}

                 />
  
           </div>

          </div> 

      </div>);


 }

}


const mapStateToProps = (state)=>{

 return {

   filters:state.filters

 };

};

export default connect(mapStateToProps)(ExpenseListFilters);