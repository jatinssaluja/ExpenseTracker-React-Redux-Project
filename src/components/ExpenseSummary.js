import React from 'react';
import {connect} from 'react-redux';
import {Link} from 'react-router-dom';
import totalExpenses from '../selectors/expenses-total';
import getVisibleExpenses from '../selectors/expenses';
import numeral from 'numeral';

export const ExpenseSummary = (props)=>{


  return (

    <div className="summary-container">

      <h2 className="summary-container__title">Viewing <span>{props.expenses.length}</span> expense(s) amounting to <span>{numeral((totalExpenses(props.expenses))/100).format('$0,0.00')}</span></h2>

       <div>

       <Link to="/create" className="btn btn-full">Add Expense</Link>

          </div>


       </div>


  	);



};



const mapStateToProps = (state)=>{


   return {

   expenses: getVisibleExpenses(state.expenses, state.filters)

   };



};









export default connect(mapStateToProps)(ExpenseSummary);