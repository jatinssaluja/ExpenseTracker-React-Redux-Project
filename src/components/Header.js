import React from 'react';
import {Link} from 'react-router-dom';



const Header = ()=>(

   <header className="header">
     
	     <Link to="/" className="header__title">
	     <h1> Expense Tracker </h1>
	     </Link>

     </header>

);

export default Header;