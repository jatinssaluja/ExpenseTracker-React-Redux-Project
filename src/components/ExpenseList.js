import React from 'react';
import {connect} from 'react-redux';
import Expense from './Expense';
import getVisibleExpenses from '../selectors/expenses';

export const ExpenseList = (props)=>{

if(props.expenses.length === 0) {

    return (

  <div className="filters-container">

    <div className="list-header">
       <div className="show-for-mobile">Expenses</div>
       <div className="show-for-desktop">Expense</div>
       <div className="show-for-desktop">Amount</div>
       </div>

      
      
 
             <p className="list-item">No Expenses</p>
          
      </div>

    );

   }
   	
 else {   

  return (<div className="filters-container">

     <div className="list-header">
       <div className="show-for-mobile">Expenses</div>
       <div className="show-for-desktop">Expense</div>
       <div className="show-for-desktop">Amount</div>
       </div>


    {props.expenses.map((expense)=>{

     
       return <Expense key = {expense.id} expense = {expense}/>;



    })}

   
    </div>



);

 }

};


const mapStateToProps = (state)=>{

 return {
 
  expenses: getVisibleExpenses(state.expenses, state.filters)

 };


};


export default connect(mapStateToProps)(ExpenseList);


