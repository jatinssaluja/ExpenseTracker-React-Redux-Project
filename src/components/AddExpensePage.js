import React, {Component} from 'react';
import ExpenseForm from './ExpenseForm';
import {connect} from 'react-redux';
import {addExpense} from '../actions/expenses';


export class AddExpensePage extends Component {

constructor(props) {
super(props);


this.onSubmit = this.onSubmit.bind(this);

}


onSubmit(expense) {

       console.log(expense);
       this.props.onSubmit(expense);
       this.props.history.push('/');


}


render() {


  return (


   <div>

    <div className="addexpense-header">

	     <h2 className="summary-container__title">Add Expense </h2>

	  </div>


     <div className="filters-container">
    <ExpenseForm  onSubmit = {this.onSubmit}/>

      </div>


    </div>  

  	);

}


}




const mapDispatchToProps = (dispatch)=>({

onSubmit: (expense)=> dispatch(addExpense(expense))



});



export default connect(undefined, mapDispatchToProps)(AddExpensePage);