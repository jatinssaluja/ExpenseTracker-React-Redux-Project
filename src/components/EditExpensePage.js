import React from 'react';
import {connect} from 'react-redux';
import ExpenseForm from './ExpenseForm';
import {editExpense, removeExpense} from '../actions/expenses';

const EditExpensePage = (props)=> {


   console.log(props.expense);

   return (<div>

        <div className="addexpense-header">

       <h2 className="summary-container__title">Edit Expense </h2>

          </div>

     <div className="filters-container">
        <ExpenseForm expense={props.expense}
                 onSubmit = {(updates)=>{

                 props.dispatch(editExpense(props.expense.id, updates));
                 props.history.push('/');

                 }}  />

      <button className="rmbtn" onClick={()=>{

          props.dispatch(removeExpense(props.expense.id));
          props.history.push('/');

      }}>Remove Expense</button>

       </div>

     </div>);


	};












const mapStateToProps = (state, props)=>{

  const expense = state.expenses.find((expense)=>{

          return expense.id === props.match.params.id;

  });

  return {

   expense

  };


};


export default connect(mapStateToProps)(EditExpensePage);