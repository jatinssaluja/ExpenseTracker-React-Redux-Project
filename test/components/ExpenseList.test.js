import {shallow} from 'enzyme';
import React from 'react';
import {ExpenseList} from '../../src/components/ExpenseList';
import toJSON from 'enzyme-to-json';
import expenses from '../fixtures/expenses';


test('should render ExpenseList with Expenses', ()=>{

const wrapper = shallow(<ExpenseList expenses={expenses}/>);
expect(toJSON(wrapper)).toMatchSnapshot();


});


test('should render ExpenseList with No Expenses message', ()=>{

const wrapper = shallow(<ExpenseList expenses={[]}/>);
expect(toJSON(wrapper)).toMatchSnapshot();


});