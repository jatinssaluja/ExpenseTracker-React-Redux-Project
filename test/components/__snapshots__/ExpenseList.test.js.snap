// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render ExpenseList with Expenses 1`] = `
<div>
  <h1>
    Expense List
  </h1>
  <Expense
    expense={
      Object {
        "amount": 195,
        "createdAt": 0,
        "description": "Gum",
        "id": "1",
        "note": "",
      }
    }
  />
  <Expense
    expense={
      Object {
        "amount": 19500,
        "createdAt": -345600000,
        "description": "Rent",
        "id": "2",
        "note": "",
      }
    }
  />
  <Expense
    expense={
      Object {
        "amount": 14500,
        "createdAt": 345600000,
        "description": "Credit Card",
        "id": "3",
        "note": "",
      }
    }
  />
</div>
`;

exports[`should render ExpenseList with No Expenses message 1`] = `
<p>
  No Expenses
</p>
`;
